### DISCUSSION TITLE
WindTerm 2.1.0

### DISCUSSION THREAD

**Discussion by kingToolbox** [kingToolbox/WindTerm#128] [General]
# Usage:
### Windows
Please install **[vc_redist.x86](https://aka.ms/vs/16/release/vc_redist.x86.exe)** if WindTerm **failed** to run.
Please use **WindTerm_Win10_Conpty.exe** to support Windows 10 shell. (**Windows 19H1 [v10.0.18309] and later**)

### MacOS
Please open dmg file and drag WindTerm.app to the `Applications` folder.

### Linux
Please use the following command to run the WindTerm after downloading.
``` shell
tar -xzvf WindTerm_2.1.0_Linux_Portable.tar.gz && cd ./WindTerm_2.1.0 && chmod +x ./WindTerm && ./WindTerm
```

### Upgrade
For Windows and Linux, copy or move the **profiles** folder from the **old Windows/Linux version** to the **new Windows/Linux version** and all the sessions and configuration will be restored.

For MacOS, copy or move the **profiles** folder from the old **WindTerm.app/Contents/MacOS** to the new **WindTerm.app/Contents/MacOS** and all the sessions and configuration will be restored.

The **profiles** folder is cross-platform. 

# Known Issues:
- [ZMODEM] **Only support file names contains ASCII characters in Windows, otherwise the file name of the transfered file will be garbled after transmission, but it does not affect the correctness of the file content.** Linux version does not have this problem.

# Features:

- [APP] **Added crash reporting system in Windows version.** #109 #110 
- [APP] **Supports changing the master password.** #70 [Intro video](https://kingtoolbox.github.io/2021/04/23/change-master-password/)
- [MacOS] **Supports MacOS and MacOS shells, e.g. bash, tch and zsh.** #44 #46 #65 
- [MacOS] **Automatically detect and add openssh, shell, putty, powershell core sessions.**
- [SESSION] **Supports synchronizing input in multiple sessions.** #87 [Intro video](https://kingtoolbox.github.io/2021/05/27/sync-input/)
- [SESSION] **Supports pausing and resuming synchronization during synchronizing input.**
- [EXPLORER] **Added a new `Explorer` pane for easier management of sessions.** [Intro video](https://kingtoolbox.github.io/2021/05/27/explorer/)
- [TAB] **Supports hovering and selecting the tab below, the default selection time is 0.2 second.** #82 
- [TAB] **Supports using `Ctrl+Alt+1` to `Ctrl+Alt+4` to select the first to fourth tab, and `Ctrl+Alt+5` to select the last tab.** #82 #86
- [TAB] Supports custom mouse double click tab operations, including `None`, `Close tab` and `Rename tab`, the default is `Close tab`. 
- [TAB] Supports custom mouse middle click tab operations, including `None`, `Close tab` and `Rename tab`, the default is `Rename tab`. 
- [TAB] Supports custom mouse right click tab operations, including `None`, `Close tab` and `Rename tab`, the default is `None` to show the tab menu. 
- [GUI] **Supports drawing text with floating point precision.**
- [GUI] **Supports focus mode.** [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_focus_mode/)
- [GUI] **Supports right-click-to-paste.** #58 #116 #117
- [GUI] **Supports specifying character spacing through `charSpaceFactor`, which range is `0.8` to `2.0`, the default is `1.0`.** #72 
- [GUI] **Supports showing tab number in the tabbar, including `None`, `All` and `FirstNine`, the default is `FirstNine`.**
- [GUI] Supports setting whether to ask to restore the last sessions again when closing the application.
- [GUI] Supports displaying the name of the current tab in the window title bar.
- [GUI] Supports keeping the window always on top.
- [GUI] Supports showing or hiding the menu bar, status bar, quick bar, etc;
- [GUI] Added a layout button in the menu bar to easily show or hide docks, panes, menu bar, status bar, etc. [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_layout/)
- [GUI] Added the right-click menu in the margins of the text view to show or hide the margins, symbols, white spaces, etc.  [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_text_margins/)
- [GUI] Make the `Scheme` buttons in the status bar clickable.
- [GUI] Supports customizing text elide mode of the tab, including `ElideLeft`, `ElideRight`, `ElideMiddle` and `ElideNone`, the default is `EchoMiddle`.
- [GUI] Supports customizing what tab should be set current when the current tab removed, including `SelectLeftTab`, `SelectRightTab`, `SelectPreviousTab`, the default is `SelectPreviousTab`.
- [GUI] **Name each tab group.**
- [GUI] Supports splitting a session tab to the specific tab group. [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_tab/)
- [GUI] Supports moving a session tab to the specific tab group. [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_tab/)
- [GUI] Supports merging a tab group into another tab group. [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_tab_group/)
- [GUI] Supports closing a tab group and all internal session tabs. [Intro video](https://kingtoolbox.github.io/2021/06/28/ui_tab_group/)
- [GUI] Supports customizing echo mode of the user name in the login wizard, , including `EchoNone`, `EchoEdit`, `EchoRandom` and `EchoAll`, the default is `EchoNone`.
- [GUI] Added a menu item `Open session` in the `Session` menu.
- [GUI] Added close buttons in the tab list menu.
- [GUI] Added a close button in the search pane.
- [GUI] Added mnemonics to the tab titles in the `Session Property` dialog.
- [GUI] Added password strength meters to the master password dialog and lock screen.
- [GUI] **Supports locking the screen, and the defaut shortcut key is `Ctrl+Shift+L`.** #70 [Intro video](https://kingtoolbox.github.io/2021/04/23/lock-screen/)
- [GUI] Make the `Local/Remote Mode` and `Lock Screen` buttons in the status bar clickable. 
- [GUI] New master password dialog with the safe input boxes.
- [TERM] Supports `OSC 7` (gnome variant) and `OSC 9;9` (ConEmu variant) working directory notifications.
- [TERM] Supports `OSC 777` (urxvt variant) and `OSC 9` (iTerm2 variant) system message notifications.
- [SHELL] Supports duplicating the working directory of a session that supports `OSC 7` or `OSC 9;9` when it is duplicated. 
- [SHELL] When dragging and dropping files or folders to the shell session, each path will be entered automatically. 
- [SENDER] Supports sending text to the specific tab group.
- [MODEM] **Supports XModem, YModem and ZModem in Telnet and Linux Shell.**
- [MODEM] **Supports specifying the use of XModem, YModem or ZModem when uploading files by drag and drop.**
- [XMODEM] **Supports using the XModem and XModem-1K protocol to download or upload one or more files not larger than 4 gigabytes.** [Intro video](https://kingtoolbox.github.io/2021/05/27/modem-xmodem/)
- [XMODEM] Supports automatic opening of the download directory after downloading the files.
- [XMODEM] Supports customizing padding char when uploading files.
- [XMODEM] Supports automatic trimming padding when downloading files.
- [YMODEM] **Supports using the YModem and YModem-1K protocol to download or upload one or more files not larger than 4 gigabytes.** [Intro video](https://kingtoolbox.github.io/2021/05/27/modem-ymodem/)
- [YMODEM] Supports automatic opening of the download directory after downloading the files.
- [YMODEM] Supports download operations such as `overwrite`, `rename` and `skip` when a local file with the same name exists.
- [ZMODEM] **Supports using the ZModem protocol to download or upload one or more files not larger than 4 gigabytes.** #27 #60  [Intro video](https://kingtoolbox.github.io/2021/05/27/modem-zmodem/)
- [ZMODEM] **Supports automatic triggering of upload and download operations.**
- [ZMODEM] **Supports automatic opening of the download directory after downloading the files.**
- [ZMODEM] **Supports upload operations such as `overwrite`, `append`, `rename` and `skip` when a remote file with the same name exists.**
- [ZMODEM] **Supports using the `Cancel` button, or `<ctrl+x><ctrl+x>`, or `<ctrl+c>` key to terminate the transmission at any time.**
- [SENDER] Supports sending escaped characters, including `\nnn`, `\xHH`, `\uHHHH`, `\c-x`, `\m-x`, `\0`, `\t`, `\r`, `\n` and so on. [Intro video](https://kingtoolbox.github.io/2020/06/20/sender-send-text/)
- [QUICKBAR] Supports sending escaped characters, including `\nnn`, `\xHH`, `\uHHHH`, `\c-x`, `\m-x`, `\0`, `\t`, `\r`, `\n` and so on. [Intro video](https://kingtoolbox.github.io/2020/08/22/quickbar/) #84 

# Improvements:
- [SSH] **Add salt to the encrypted login information makes it possible to generate different encryption results even when the same login information is encrypted to enchance security.** 
- [SSH] **Clear the decrypted login information from the memory immediately after completing the automatic login to enhance security.**
- [SSH] When the saved automatic authentication is cleared, useless configuration items `ssh.port` and `ssh.target` are no longer generated in the configuration file.
- [APP] Enhance the verification of the correctness of the master password. #85 
- [GUI] Make the opened sessions displayed first in the `Open Session` dialog.
- [GUI] Simplify the trigger of `Show All Opened and Recent Terminals` in the `Command Palette` from `tm` to `t`.
- [GUI] Supports setting the `default editor` path to an empty string to clear it. 
- [GUI] When the window gets the focus and no session is opened, set the focus to the address bar to facilitate input. #120
- [GUI] Darken the scrollbar slider color in the themes `dige-black-and-white` and `dige-white` to make it more eye-catching. #81 
- [GUI] Avoid session views that can be resized down to size 0.
- [GUI] Add more detailed information to the session tab tooltip, including `Session`, `Target`, `Tab Index`, `Sync Channel`, `Tab Title` and `Tab Group`. 
- [GUI] The tooltip of the session tab displays more tab page information, including 
- [GUI] Adjust the dige-black, dige-white and dige-black-and-white color themes for better display.
- [GUI] Redesign the master password dialog box and add instruction descriptions. 
- [GUI] Added tooltips to the message bar.
- [GUI] Align menu item icons and check marks.
- [GUI] Added a count label to the filter input box of the tab list menu.
- [GUI] The tooltip of the session tab displays more tab page information, including `Session`, `Target`, `Tab Index` and `Sync Channel`.
- [TAB] Supports asking whether to close the current session when double clicking the session tab.
- [FILER] The file size displayed in the status bar is accurate to 2 digit precision, like 1.42GB.
- [SENDER] Added icons and tooltips to the session selection list.
- [SCHEME] Supports matching the number of milliseconds in the datetime. 

# Bugs:
- [APP] **The application may crash when opening Windows shells, like `Cmd` and `Powershell`.** #110 
- [APP] **The application may crash when setting the number of columns and rows of the windows to large numbers.** #126 
- [APP] Splitting the tabgroup multiple times will cause a crash when restoring the layout after restarting. 
- [APP] **When the automatic login information of any session is cleared, the master password will become invalid.** #105 
- [APP] Cannot restart after exiting in full screen mode. #93 #99 
- [LOG] Cannot record logs in the binary mode of the Serial and raw TCP. #71 
- [LOG] When recording the log, the long line is recorded in segments. #126
- [GUI] The current zoom setting of the font will be reset when switching the theme. 
- [GUI] The text widget in the `Sender` pane cannot correctly respond to the shortcut key `Alt+O`. 
- [GUI] Window failed to properly restore to maximum when exiting full screen. #79 
- [GUI] The new tabgroup does not get focus when the old tabgroup is closed. 
- [GUI] The text is not displayed properly when the screen is changed. #73 #78 
- [GUI] When the `Find` combo box contains a long string, the width will be too wide, even beyond the screen. #83 
- [GUI] When the text encoding of the server is not `UTF-8`, and the window title set by the server contains non-ASCII characters, the window title will be garbled.
- [GUI] When using the menu item `View - Panes - Filer` to activate the filer pane, the filer pane cannot accept keyboard input immediately. 
- [GUI] The text is not displayed properly when the screen resolution is changed. #72
- [GUI] When the shell combobox in the session dialog is switched to admin:cmd, admin:powershell, gitbash, cygwin, etc., the system combobox fails to automatically set the corresponding system type. 
- [SSH] **The session cannot log in automatically when using pageant authentication.** #70 
- [TERM] The shortcut key `Ctrl+@` cannot send `NUL`. #123
- [TERM] The key `AltGr` is erroneously handled as key `Del`. #121 
- [SENDER] The shortcut key `<Alt+Enter>` does not start sending text. #87  
- [SENDER] Cannot respond to any keys except the `<Space>` key in the sessions list.

<hr /><em>This discussion was created from the release <a href='https://github.com/kingToolbox/WindTerm/releases/tag/2.1'>WindTerm 2.1.0</a>.</em>

**Comment by danmur** [comment]
w00t, the crash I was getting under windows seems to be fixed :)

**Reply by kingToolbox** [reply]
Yes, the stable version naturally needs to be more stable to be worthy of the name. :smile: And this is also inseparable from everyone's help, thank you all.

**Comment by gooker** [comment]
You did a great job.

**Reply by kingToolbox** [reply]
Thank you, my pleasure.

**Comment by Twink-Polar** [comment]
Congratulations on the 2.1.0 release!

**Reply by kingToolbox** [reply]
Thank you and thank you for your continued support.

**Comment by itInner** [comment]
Great job!!! Thank you for developing  this amazing tool.
2.1.0 suooprts  ZModem，so  I can use lrzsz upload and download file easer then before.
A little suggestion:
May I buy you a coffee?

**Reply by kingToolbox** [reply]
Thank you very much, and I am very very grateful for your generosity. Please feel free to use WindTerm, I am very happy that you like WindTerm, and your praise is already my greatest reward. :smile:
